### DISCUSSION TITLE
Tailwind escaped classes are not purged

### DISCUSSION THREAD

**Discussion by aadilmehrajbhat** [tailwindlabs/tailwindcss#4778] [Help]
### What version of Tailwind CSS are you using?

v2.0.1

### What build tool (or framework if it abstracts the build tool) are you using?

postcss-cli 8.3.0, tailwindcss 2.0.1, postcss 8.1.14, autoprefixer 10.0.4

### What version of Node.js are you using?

v12.22.1

### What browser are you using?

Chrome

### What operating system are you using?

Ubuntu

### Reproduction repository

https://github.com/AadilMehrajBhat/tailwindcss-purge

### Describe your issue

I have a use case in which I need to persist previous build CSS for backward compatibility, in order to achieve the same I was adding the previous bundle CSS as one of the purge path entries to retain those tailwind classes, which worked very well. But for the tailwind classes that are escaped,  are not getting added in the resulting bundle.

E.g.
Previous CSS bundle:
```
.text-sm{font-size:.875rem;line-height:1.25rem}

.md\:order-last{order:9999}
```

When I add this file in the purge path:
```
 purge: ['./dist/**/*.css', ...]
```

The resulting bundle only includes the not escaped classes and ignores the escaped class, like this:
```
.text-sm{font-size:.875rem;line-height:1.25rem}
```

After I change `.md\:order-last` to `.md:order-last` in the original file, it is being added to the resulting bundle:
```
@media (min-width: 768px) {
  .md\:order-last {
    order: 9999
  }
}
```

Is there any way by which I can unescape the CSS classes while the content gets purged?

**Comment by sachinraja** [comment]
> I have a use case in which I need to persist previous build CSS for backward compatibility

If you don't mind me asking, what is that use case?

**Comment by aadilmehrajbhat** [comment]
> > I have a use case in which I need to persist previous build CSS for backward compatibility
> 
> If you don't mind me asking, what is that use case?

Well there are HTML pages that are already In the system (database) with the old classes and all of them share the same CSS file,  for that reason I need to persist already used classes for the backward compatibility.

**Comment by sachinraja** [comment]
How did you generate the old CSS build file?

**Comment by adamwathan** [comment]
Hey! This isn't really a bug, it's working correctly because PurgeCSS expects to be looking at HTML, not at CSS, so it thinks the classes with escaped characters just contain slashes, like if they were in HTML like this:

```html
<div class="md\:order-last">
```

I would recommend just keeping the old CSS file and not running PurgeCSS against that one at all, or using their control comments to preserve parts of your CSS:

https://purgecss.com/safelisting.html#in-the-css-directly

Going to convert this to a discussion for you since it's not a bug in Tailwind or anything 👍

**Reply by aadilmehrajbhat** [reply]
Thank you @adamwathan for the prompt reply!

Is there a way by which we can transform the CSS to unescape before getting purged? If not can you please point me to the code where tailwind escapes the classes so that I can reverse it manually,

**Comment by antoinerey** [comment]
I'm also looking for a solution to this.

---

As a bit of context:

- I'm using Tailwind to develop a design system in a repository (A). This repository contains the configuration (plugins, theme…), and outputs Vue components and the generated CSS file that only includes classes used by those components.

- Then, I'm consuming that design system from another repository (B). Because I want to keep them in sync and share the same theme, I'm importing Tailwind configuration from A into B, and generate a new CSS file. However, I only want to load a single CSS file that would include classes from A _and_ B, so I'm looking for a way to take into account classes generated from A.

--- 

For now, I've tried:

- Including the generated CSS file from A to the `purge.content` array. (1)

  ```js
  module.exports = {
    purge: {
      content: [
        'src/**/*.vue',
        'node_modules/design-system/dist/styles.css',
      ],
    },
  }
  ```

- Transforming the CSS file before looking for classes. (2)

  ```js
  module.exports = {
    purge: {
      content: [
        'src/**/*.vue',
        'node_modules/design-system/dist/styles.css',
      ],
    },

    options: {
      transform: {
        css: (content) => {
          return content.replaceAll('\\', '')
        }
      },
    },
  }
  ```

- Registering the CSS file under `purge.options.css`. (3)

  ```js
  module.exports = {
    purge: {
      content: [
        'src/**/*.vue',
      ],

      options: {
        css: [
          'node_modules/design-system/dist/styles.css',
        ],
      },
    },
  }
  ```

None of these solutions worked unfortunately. 

- 1 and 2 do take into account non-escaped classes, but fail when processing escaped classes such as `.w-\[90px\]`.
- 3 does not include any classes from the design system at all.

I'm starting to run out of ideas here. Any help is welcome. 🙏

**Reply by aadilmehrajbhat** [reply]
@antoinerey I ended up writing a script which uses RegEx to unescape the classes right before purging, so that resulting bundle will include those classes as well. That worked for me

**Reply by antoinerey** [reply]
@AadilMehrajBhat Would you mind sharing an overview of how you configured Tailwind to do that? 🙏

**Reply by sachinraja** [reply]
I haven't tried it, but you could try using [gulp-concat-css](https://www.npmjs.com/package/gulp-concat-css) and purgecss. Or you could run postcss on both files, concatenate them, and then run [cssnano](https://www.npmjs.com/package/cssnano) to remove the duplicates. There's also [css-purge](https://github.com/rbtech/css-purge), which may do all of this for you.

**Reply by aadilmehrajbhat** [reply]
@antoinerey I wrote a custom node script which would replace the escaped tokens and then using `yargs` ran the tailwind build command.

Something like this:
```
function escapeExistingCssBundle({ bundleCssPath, bundleCss }) {
  const DELIMETER = '\\';
  const ESCAPE_TOKENS = new Set(['.', ':', '/', '3'].map((t) => DELIMETER + t));
  const tokenRegex = /\\(.)/gm;

// In case a new escaped token was found not present in the available token list, then identify
// that token and update ESCAPE_TOKENS  manually. 
  function hasIdenticalTokens(tokens) {
    for (let token of tokens) {
      if (!ESCAPE_TOKENS.has(token)) {
        console.log(
          `'\\${token}' not found in the list of tokens`,
          ESCAPE_TOKENS,
        );
        return false;
      }
    }
    return true;
  }
  const tokens = new Set(bundleCss.match(tokenRegex));
  if (!tokens.size || !hasIdenticalTokens(tokens)) return;

  const result = bundleCss.replace(tokenRegex, '$1');

  fs.writeFileSync(bundleCssPath, result);
}
```

**Reply by antoinerey** [reply]
Ah thanks a lot! I'll definitively take a look and maybe get inspired.

I'm a bit confused though. Is it not the role of the `options.transform` to take care of these transformations before the purge happens?
